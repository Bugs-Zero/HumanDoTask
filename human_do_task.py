from __future__ import annotations

import datetime
import traceback
import typing


class Clock(object):
    started: typing.Optional[datetime.datetime]

    def __init__(self) -> None:
        self.reset()

    def reset(self) -> None:
        self.accumulator = datetime.timedelta(0)
        self.started = None

    def start(self) -> None:
        if not self.started:
            self.started = datetime.datetime.utcnow()

    def stop(self) -> None:
        if self.started:
            self.accumulator += datetime.datetime.utcnow() - self.started
            self.started = None

    @property
    def elapsed(self) -> datetime.timedelta:
        if self.started:
            return self.accumulator + (datetime.datetime.utcnow() - self.started)
        return self.accumulator

    def __repr__(self) -> str:
        return "<Clock {} ({})>".format(
            self.elapsed, "started" if self.started else "stopped"
        )


def _query(
    prefix: list[str], question: str, first_answer: str, second_answer: str
) -> bool:
    for item in prefix:
        print(item)
    return (
        input(f"{question} ({first_answer}/{second_answer})")
        .lower()
        .startswith(first_answer.lower())
    )


class ManualSection:
    def __init__(self, go: Process) -> None:
        self.go = go

    def __enter__(self) -> None:
        self.go._automated.stop()
        self.go._manual.start()

    def __exit__(
        self,
        exception_type: typing.Optional[type[BaseException]],
        exception: typing.Optional[BaseException],
        exception_traceback: typing.Optional[traceback.TracebackException],
    ) -> None:
        self.go._manual.stop()
        self.go._automated.start()


class Process:
    def __init__(self) -> None:
        self._automated = Clock()
        self._automated.start()
        self._manual = Clock()
        self._test_result: list[str] = []
        self._automated_steps = 0
        self._manual_steps = 0

    @staticmethod
    def run(
        perform: typing.Callable[[Process], None],
        verify: typing.Callable[[Process], None],
    ) -> None:
        go = Process()
        with ManualSection(go):
            do_perform = _query(
                [], "Do you wish to perform the process or verify it?", "P", "V"
            )
        if do_perform:
            perform(go)
        else:
            verify(go)
        go._print_stats()

    def do(self, operation: typing.Callable[[], None]) -> None:
        self._automated_steps += 1
        operation()

    def tell(self, message: str) -> None:
        self._manual_steps += 1
        with ManualSection(self):
            print(message)
            input("press enter when done")

    def ask(self, condition: str, operation: typing.Callable[[], None]) -> None:
        self._manual_steps += 1
        with ManualSection(self):
            should_do_it = _query([condition], "Should I perform this step?", "Y", "N")
        if should_do_it:
            operation()

    def ask_yes_no(self, condition: str) -> bool:
        self._manual_steps += 1
        with ManualSection(self):
            return _query([condition], "Should I perform this step?", "Y", "N")

    # Conflicts with built-in keyword `if`. TODO: pick a non-conflicting name.
    #    def if(self, condition, operation) -> None:
    #       self._automatic_steps += 1
    #       if(condition()):
    #          operation()

    def verify(self, condition: typing.Callable[[], bool]) -> None:
        initial = self._manual_steps
        if not condition():
            self._test_result.append(f"Failed expectation: {condition}")
        if initial == self._manual_steps:
            self._automated_steps += 1

    def that(self, condition: str) -> typing.Callable[[], bool]:
        def impl() -> bool:
            self._manual_steps += 1
            with ManualSection(self):
                return _query(
                    [f"Please verify whether {condition}."], "Is this right?", "Y", "N"
                )

        return impl

    def print_test_results(self) -> None:
        if self._test_result:
            print("Verification failed. Please fix the process and try again.")
        for failure in self._test_result:
            print(failure)

    def _print_stats(self) -> None:
        self._automated.stop()
        total_time = self._automated.elapsed + self._manual.elapsed
        print(f"Process complete in {total_time}.")
        print(
            f"   Automated: {self._automated_steps} steps in {self._automated.elapsed}."
        )
        print(f"   Manual: {self._manual_steps} steps in {self._manual.elapsed}.")
