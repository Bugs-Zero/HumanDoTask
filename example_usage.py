from human_do_task import Process

# begin-snippet: example_usage
def perform(go: Process) -> None:
    go.tell("Please lorem.")
    go.tell("Please ipsum.")
    go.ask(
        "Are the lights on?", lambda: go.tell("Please turn off the circuit breaker.")
    )
    go.tell("Please do one more thing.")


def verify(go: Process) -> None:
    go.tell("Please do something.")
    go.verify(go.that("the thing is colored blue"))
    # end-snippet


if __name__ == "__main__":
    Process.run(perform, verify)
