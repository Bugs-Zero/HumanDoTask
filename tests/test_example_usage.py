import re
import textwrap
import traceback
import typing
import approvaltests
import approvaltests.approvals
import contextlib
import io
import unittest
import unittest.mock

import example_usage
import human_do_task


def _scrub_timestamp(s: str) -> str:
    return re.sub(r"\d+:\d+:\d+(\.\d+)?", "{REMOVED}", s)


class redirect_stdin(contextlib._RedirectStream[io.StringIO]):
    _stream = "stdin"


@contextlib.contextmanager
def capture_io(
    input: str,
) -> typing.Generator[io.StringIO, None, None]:
    with contextlib.redirect_stdout(io.StringIO()) as stdout:
        with redirect_stdin(io.StringIO(input)):
            yield stdout


def test__verify_happy_path__run() -> None:
    with capture_io(
        textwrap.dedent(
            """\
            P


            N


            """
        )
    ) as stdout:
        human_do_task.Process.run(example_usage.perform, example_usage.verify)

    approvaltests.approvals.verify(
        stdout.getvalue(),
        options=approvaltests.Options().with_scrubber(_scrub_timestamp),
    )


def test__verify_happy_path__verify() -> None:
    with capture_io(
        textwrap.dedent(
            """\
        V

        Y

        """
        )
    ) as stdout:
        human_do_task.Process.run(example_usage.perform, example_usage.verify)

    approvaltests.approvals.verify(
        stdout.getvalue(),
        options=approvaltests.Options().with_scrubber(_scrub_timestamp),
    )
