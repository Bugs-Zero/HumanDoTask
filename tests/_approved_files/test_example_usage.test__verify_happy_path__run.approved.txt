
Please lorem.

Please ipsum.

Are the lights on?

Please do one more thing.

Process complete in {REMOVED}.
   Automated: 0 steps in {REMOVED}.
   Manual: 4 steps in {REMOVED}.
